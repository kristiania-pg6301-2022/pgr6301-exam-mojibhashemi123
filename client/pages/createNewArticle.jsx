import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { articleContext } from "../context/articleContext";

export function CreateNewArticle({ user }) {
  const { createArticle } = useContext(articleContext);
  const [title, setTitle] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [text, setText] = useState("");
  const [username, setUsername] = useState(user.microsoft?.name);
  const [error, setError] = useState();

  const navigate = useNavigate();

  async function handleSubmit(event) {
    event.preventDefault();
    const result = createArticle({
      title,
      author,
      topic,
      text,
      username,
    });

    try {
      if (title === "" || author === "" || topic === "" || text === "") {
        setError("Du må fylle");
        return;
      }
      await result;
      navigate("/");
    } catch (error) {
      console.error("ERROR: ", error);
      setError("Du er ikke pålogget");
    }
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h1>Add new article</h1>
        <FormInput label={"Title:"} value={title} onChangeValue={setTitle} />
        <FormInput label={"Topics:"} value={topic} onChangeValue={setTopic} />
        <FormInput label={"Author:"} value={author} onChangeValue={setAuthor} />
        <strong>Text area:</strong>
        <textarea value={text} onChange={(e) => setText(e.target.value)} />
        {error && <div>{error}</div>}
        <div>
          <button>Submit</button>
        </div>
      </form>
    </div>
  );
}
