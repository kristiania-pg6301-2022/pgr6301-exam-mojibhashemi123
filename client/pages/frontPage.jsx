import React from "react";
import { Link } from "react-router-dom";

export function FrontPage() {
  return (
    <div id="frontpage">
      <h1>news article</h1>
      <ul>
        <li>
          <Link to={"/"}>Show news article</Link>
        </li>
        <li>
          <Link to={"/article/new"}>Add new article</Link>
        </li>
        <li>
          <Link to={"/article/update"}>Update your Article</Link>
        </li>
      </ul>
    </div>
  );
}
