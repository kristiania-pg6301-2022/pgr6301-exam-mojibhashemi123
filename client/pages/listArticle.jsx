import { useContext, useEffect, useState } from "react";
import { articleContext } from "../context/articleContext";
import { useLoading } from "../useLoading";
import { Link, Route, Routes } from "react-router-dom";
let articles = [];
export function ListArticle() {
  const { fetchArticle } = useContext(articleContext);

  const { data, error, loading, reload } = useLoading(fetchArticle);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  function handleClick() {
    document.getElementById("sidebar").classList.toggle("active");
  }

  function ListSideBar(event) {
    document.getElementById("sidebar").classList.toggle("dd");
    if (event) {
      console.log(event);

      articles = [];

      data.map((article, index) => {
        if (article.topic === event) {
          document.getElementById("sidebar").classList.toggle("dd");
          console.log(article.title + " fra sidebar");
          articles.push(article);
          //return ShowMatch(articles);
        }
      });
    }
  }
  let listTopic = Array.from(new Set(data.map((t) => t.topic)));

  return (
    <div id="#container2">
      <div id="sidebar" onClick={({ target }) => handleClick(target)}>
        <h1>Articles</h1>
        <div className={"toggle-btn"}>
          <span></span>
          <span></span>
          <span></span>
        </div>
        <ul>
          {listTopic.map((topic, index) => (
            <li onClick={({ target }) => ListSideBar(topic)}>
              <Link to={"/ShowMatch"}>{topic}</Link>
            </li>
          ))}
        </ul>
      </div>
      <h1></h1>
    </div>
  );
}

export function ShowMatch() {
  return (
    <div id="container2">
      {articles.map((articles, index) => (
        //for feil meldinger i browser bruk index
        <ArticleCard key={index} articles={articles} />
      ))}
    </div>
  );
}

function ArticleCard({ articles: { title, author, topic, text, date } }) {
  return (
    <>
      <h3>Title: {title}</h3>
      <h4>Author: {author}</h4>
      <div>Topic: {topic}</div>
      <p>Text: {text}</p>
      <h6>Date: {date}</h6>
      <p>
        -------------------------------------------------------------------------------------------------------------------------------------------------------------
      </p>
    </>
  );
}
