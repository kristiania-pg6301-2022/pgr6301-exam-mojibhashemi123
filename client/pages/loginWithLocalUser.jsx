import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { UserContext } from "../context/userContext";
import { FormInput } from "../lib/formInput";

import "../css/loginWithLocalUser.css";

export function LoginWithLocalUser() {
  const { loginSession } = useContext(UserContext);
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState();

  const navigate = useNavigate();

  async function handleSubmit(event) {
    event.preventDefault();
    const result = loginSession({ email, password });

    try {
      await result;
      navigate("/");
    } catch (error) {
      console.error("ERROR: ", error);
      setError("Du har tastet feil eller m√• du registerere deg");
    }
  }

  return (
    <div className={"loginDiv"}>
      <form onSubmit={handleSubmit}>
        <FormInput label={"Email:"} value={email} onChangeValue={setEmail} />
        <FormInput
          label={"Password:"}
          value={password}
          onChangeValue={setPassword}
        />
        {error && <div>{error}</div>}
        <div>
          <button>Log in</button>
        </div>
      </form>
    </div>
  );
}
