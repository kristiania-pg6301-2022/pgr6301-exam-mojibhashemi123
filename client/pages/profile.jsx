import React from "react";
export function Profile({ user }) {
  if (user.google) {
    return (
      <div className={"card"}>
        <h1>You are signed with google account</h1>
        <ShowProfile
          name={user.google.name}
          picture={user.google.picture}
          email={user.google.email}
        />
      </div>
    );
  } else if (user.microsoft) {
    return (
      <div className={"card"}>
        <h1>You are signed with school account</h1>
        <ShowProfile
          name={user.microsoft.name}
          picture={user.microsoft.picture}
          email={user.microsoft.email}
        />
      </div>
    );
  } else if (user.email) {
    const name = user.email.map((u) => u.name);
    const email = user.email.map((u) => u.email);
    return (
      <div className={"card"}>
        <h1>You are signed with local account</h1>
        <ShowProfile name={name} email={email} />
      </div>
    );
  }
}

function ShowProfile({ name, picture, email }) {
  return (
    <main>
      <div className="row">
        <div className="column">
          <div className="card">
            <img src={picture} alt={"Profile picture"} />
          </div>
          <h3>Name: {name}</h3>
          <h4>Email: {email}</h4>
        </div>
      </div>
    </main>
  );
}
