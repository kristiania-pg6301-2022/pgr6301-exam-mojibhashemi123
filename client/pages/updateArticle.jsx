import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { articleContext } from "../context/articleContext";

export function UpdateArticle({ user, reload }) {
  const { updateArticle } = useContext(articleContext);
  const [title, setTitle] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [text, setText] = useState("");
  const [username, setUsername] = useState(user.microsoft?.name);
  const [error, setError] = useState();

  const navigate = useNavigate();

  async function handleSubmit(event) {
    event.preventDefault();
    const result = updateArticle({
      title,
      author,
      topic,
      text,
      username,
    });

    try {
      if (title === "" || author === "" || topic === "" || text === "") {
        setError("Du m√• fylle");
        return;
      }
      await result;
      reload();
      navigate("/");
    } catch (error) {
      console.error("ERROR: ", error);
      setError("Bare han som har skrevet articlen kan endre");
    }
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h1>Update your article (Only School account)</h1>
        <FormInput
          label={"which article? Title:"}
          value={title}
          onChangeValue={setTitle}
        />
        <FormInput label={"Topics:"} value={topic} onChangeValue={setTopic} />
        <FormInput label={"Author:"} value={author} onChangeValue={setAuthor} />
        <strong>Text area:</strong>
        <textarea value={text} onChange={(e) => setText(e.target.value)} />
        {error && <div>{error}</div>}
        <div>
          <button>Submit</button>
        </div>
      </form>
    </div>
  );
}
