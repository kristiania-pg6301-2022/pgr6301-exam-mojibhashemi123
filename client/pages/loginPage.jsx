import React, { useContext, useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate, useParams } from "react-router-dom";
import { APIsContext } from "../apisContext";
import { randomString } from "../lib/randomString";
import { sha256 } from "../lib/sha256";
import { LoginWithLocalUser } from "./loginWithLocalUser";

export function LoginCallback({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(APIsContext);
  useEffect(async () => {
    const { access_token, error, error_description, state, code } =
      Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1))
      );

    const expected_state = window.sessionStorage.getItem("expected_state");
    if (!state || expected_state !== state) {
      setError("Unexpected state");
      return;
    }

    if (error || error_description) {
      setError(`Error: ${error} (${error_description})`);
      return;
    }

    if (code) {
      const { client_id, token_endpoint } = config[provider];
      const code_verifier = window.sessionStorage.getItem("code_verifier");

      let hostname =
        window.location.protocol +
        "//" +
        window.location.hostname +
        (window.location.port == "" ? "" : ":") +
        window.location.port;

      const payload = {
        grant_type: "authorization_code",
        code,
        client_id,
        redirect_uri: hostname + "/login/microsoft/callback",
        code_verifier,
      };
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      if (!res.ok) {
        setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
        return;
      }
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });
      reload();
      navigate("/");
      return;
    }

    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    await registerLogin(provider, { access_token });
    reload();
    navigate("/");
  }, []);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return <h1>Please wait...</h1>;
}

export function EndSession({ reload }) {
  const navigate = useNavigate();
  const { endSession } = useContext(APIsContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });
  return <h1>Please wait...</h1>;
}

function LoginButton({ config, label, provider }) {
  async function handleLogin() {
    const {
      authorization_endpoint,
      response_type,
      scope,
      client_id,
      code_challenge_method,
    } = config[provider];

    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);

    const parameters = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri: `${window.location.origin}/login/${provider}/callback`,
    };

    if (code_challenge_method) {
      const code_verifier = randomString(50);
      window.sessionStorage.setItem("code_verifier", code_verifier);
      parameters.code_challenge_method = code_challenge_method;
      parameters.code_challenge = await sha256(code_verifier);
    }

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(parameters);
  }

  return (
    <div>
      <button onClick={handleLogin}>{label}</button>
    </div>
  );
}

export function StartLogin({ config, reload }) {
  return (
    <div>
      <h1>Choice login</h1>
      <Link to={"/account/new"}>
        <h1>Sign up</h1>
      </Link>
      <LoginWithLocalUser reload={reload} />
      <LoginButton
        label={"Login with Google"}
        config={config}
        provider={"google"}
      />
      <LoginButton
        label={"Login with school account"}
        config={config}
        provider={"microsoft"}
      />
    </div>
  );
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route
        path={"/"}
        element={<StartLogin config={config} reload={reload} />}
      />
      <Route
        path={"/:provider/callback"}
        element={<LoginCallback config={config} reload={reload} />}
      />
      <Route path={"/endsession"} element={<EndSession reload={reload} />} />
      <Route path={"*"} element={<StartLogin config={config} />} />
    </Routes>
  );
}
