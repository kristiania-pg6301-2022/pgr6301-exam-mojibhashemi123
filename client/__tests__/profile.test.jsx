import ReactDOM from "react-dom";
import React from "react";
import { Profile } from "../pages/profile";
import { MemoryRouter } from "react-router-dom";

describe("profile", () => {
  it("show profile for google", () => {
    const user = {
      user: {
        google: {
          name: "NameTest",
          picture:
            "https://lh3.googleusercontent.com/a-/AOh14Ghh70gSuO_xtuhlK-4jZkflx-T6oHibukhQX7ul=s96-c",
          email: "test@test.com",
        },
      },
    };

    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Profile user={user?.user} />
      </MemoryRouter>,
      domElement
    );

    expect(
      Array.from(domElement.querySelectorAll("h1")).map((e) => e.innerHTML)
    ).toEqual(["You are signed with google account"]);

    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("show profile for microsoft", () => {
    const user = {
      user: {
        microsoft: {
          name: "testName2",
          picture:
            "https://lh3.googleusercontent.com/a-/AOh14Ghh70gSuO_xtuhlK-4jZkflx-T6oHibukhQX7ul=s96-c",
          email: "test@hotmail.com",
        },
      },
    };

    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Profile user={user?.user} />
      </MemoryRouter>,
      domElement
    );

    expect(
      Array.from(domElement.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual(["Name: testName2"]);

    expect(domElement.innerHTML).toMatchSnapshot();
  });
});
