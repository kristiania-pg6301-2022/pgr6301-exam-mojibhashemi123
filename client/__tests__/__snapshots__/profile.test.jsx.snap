// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile show profile for google 1`] = `"<div class=\\"card\\"><h1>You are signed with google account</h1><main><div class=\\"row\\"><div class=\\"column\\"><div class=\\"card\\"><img src=\\"https://lh3.googleusercontent.com/a-/AOh14Ghh70gSuO_xtuhlK-4jZkflx-T6oHibukhQX7ul=s96-c\\" alt=\\"Profile picture\\"></div><h3>Name: NameTest</h3><h4>Email: test@test.com</h4></div></div></main></div>"`;

exports[`profile show profile for microsoft 1`] = `"<div class=\\"card\\"><h1>You are signed with school account</h1><main><div class=\\"row\\"><div class=\\"column\\"><div class=\\"card\\"><img src=\\"https://lh3.googleusercontent.com/a-/AOh14Ghh70gSuO_xtuhlK-4jZkflx-T6oHibukhQX7ul=s96-c\\" alt=\\"Profile picture\\"></div><h3>Name: testName2</h3><h4>Email: test@hotmail.com</h4></div></div></main></div>"`;
